src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -14,35 +14,289 @@ namespace Microsoft.CodeAnalysis
+
+            if (referencedProjectIds.Count == 0)
+            {
+                return this;
+            }
+
+            var newReferencesMap = ComputeNewReferencesMapForAdditionalProjectReferences(_referencesMap, projectId, referencedProjectIds);
+
+            var newReverseReferencesMap =
+                _lazyReverseReferencesMap != null
+                    ? ComputeNewReverseReferencesMapForAdditionalProjectReferences(_lazyReverseReferencesMap, projectId, referencedProjectIds)
+                    : null;
+
+            var newTransitiveReferencesMap = ComputeNewTransitiveReferencesMapForAdditionalProjectReferences(_transitiveReferencesMap, projectId, referencedProjectIds);
+
+            var newReverseTransitiveReferencesMap = ComputeNewReverseTransitiveReferencesMapForAdditionalProjectReferences(_reverseTransitiveReferencesMap, projectId, referencedProjectIds);
+
+            // Note: rather than updating our dependency sets and topologically sorted data, we'll throw that away since incremental update is
+            // tricky, and those are rarely used.
+            return new ProjectDependencyGraph(
+                _projectIds,
+                referencesMap: newReferencesMap,
+                reverseReferencesMap: newReverseReferencesMap,
+                transitiveReferencesMap: newTransitiveReferencesMap,
+                reverseTransitiveReferencesMap: newReverseTransitiveReferencesMap,
+                topologicallySortedProjects: default,
+                dependencySets: default);
+        }
+
+        /// <summary>
+        /// Computes a new <see cref="_referencesMap"/> for the addition of additional project references.
+        /// </summary>
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReferencesMapForAdditionalProjectReferences(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingReferencesMap,
+            ProjectId projectId,
+            IReadOnlyList<ProjectId> referencedProjectIds)
+        {
+            if (existingReferencesMap.TryGetValue(projectId, out var existingReferences))
+            {
+                return existingReferencesMap.SetItem(projectId, existingReferences.Union(referencedProjectIds));
+            }
+            else
+            {
+                return existingReferencesMap.SetItem(projectId, referencedProjectIds.ToImmutableHashSet());
+            }
+        }
+
+        /// <summary>
+        /// Computes a new <see cref="_lazyReverseReferencesMap"/> for the addition of additional project references. Must be called on a non-null
+        /// map.
+        /// </summary>
+        private static ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReverseReferencesMapForAdditionalProjectReferences(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> existingReverseReferencesMap,
+            ProjectId projectId,
+            IReadOnlyList<ProjectId> referencedProjectIds)
```

Comment:
I find it somewhat difficult to remember which references or sets of references are pre-existing / newly added / mid-calculation / etc. when reading the algorithms themselves. Maybe "AddedReferences" or "NewDirectlyReferencedProjectIds" or something?